default/python/AI/ProductionAI.py

```diff
@@ -1812,3 +1724,204 @@ def get_number_of_existing_outpost_and_colony_ships() -> int:
+        if not building_type.can_be_produced(pid) or not building_type.can_be_enqueued(pid):
+            continue
+        if pid in opinion.likes:
+            preferred_locations.append((planet.currentMeterValue(fo.meterType.maxTroops), pid))
+        else:
+            locations.append((planet.currentMeterValue(fo.meterType.maxTroops), pid))
+    for _, pid in sorted(preferred_locations) + sorted(locations):
+        planet = universe.getPlanet(pid)
+        res = building_type.enqueue(pid)
+        debug("Enqueueing %s at planet %d (%s) , with result %d", building_type, pid, planet.name, res)
+        if res:
+            if at_front:
+                res = fo.issueRequeueProductionOrder(empire.productionQueue.size - 1, 0)  # move to front
+                debug("Requeueing %s to front of build queue, with result %d", building_type, res)
+            return building_type.turn_cost(pid)
+    return 0.0
+
+
+def _may_enqueue_for_stability(building_type: BuildingType, new_turn_cost: float) -> float:
+    """
+    Build building if it seems worth doing so to increase stability.
+    Only builds of locations.planets_enqueued is empty and new_turn_cost is 0.0,
+    i.e. there are currently no build queue entries for the given building.
+    returns new_turn_cost or turn_cost of the building enqueued by this function.
+    """
+    if building_type.queued_in() or new_turn_cost:
+        return new_turn_cost
+    # this can be improved a lot, taking into account value of planets, actual stability and
+    # what effects the change would have. For the moment, keep it simple.
+    # Note that the strongest effect is always on the building's planet itself.
+    opinion = building_type.get_opinions()
+    universe = fo.getUniverse()
+    if len(opinion.likes) >= len(opinion.dislikes):
+        like_candidates = opinion.likes - building_type.built_at(True)
+        # plans may change, so consider only actual colonies that like it
+        candidates = [pid for pid in like_candidates if universe.getPlanet(pid).speciesName]
+        return _try_enqueue(building_type, candidates)
+    return 0.0
+
+
+def _build_scanning_facility() -> float:
+    """Consider building Scanning Facilities"""
+    building_type = BuildingType.SCANNING_FACILITY
+    empire = fo.getEmpire()
+    if not building_type.available():
+        return 0.0
+
+    universe = fo.getUniverse()
+    turn_cost = 0.0
+    opinion = building_type.get_opinions()
+    # TBD use actual cost?
+    max_scanner_builds = max(1, int(empire.productionPoints / 30)) - len(building_type.queued_in())
+    scanner_systems = building_type.built_at_sys(True)
+    debug(
+        "Considering building %s, found current and queued systems %s, planets that like it %s, #dislikes: %d",
+        building_type,
+        PlanetUtilsAI.sys_name_ids(scanner_systems),
+        PlanetUtilsAI.sys_name_ids(opinion.likes),
+        len(opinion.dislikes),
+    )
+    for sys_id in get_owned_planets().keys():
```

Comment:
```suggestion
    for sys_id in get_owned_planets():
```

Dict is iterated by keys, you don't need explicitly call it here.